### ISSUE TITLE
Session close should call handlers in reverse order

### ISSUE THREAD

**Issue by msharov** [linux-pam/linux-pam#36]
Currently handlers are called in the same order for both open session and close session cases. This causes problems when one handler requires another.

For example, pam_mount handler mounts the user home directory, followed by pam_systemd handler launching user services that read and keep open files in the home directory. At close session, pam_mount is called first to unmount the home directory, and is unable to do so because pam_systemd has not stopped its services.

To fix, _add_pam_handler in libpam/pam_handlers.c needs to add the handler to the beginning of handler_p2 when type is PAM_T_SESS, rather than at the end as it currently does.

**Comment by t8m** [comment]
Unfortunately changing this would most possibly break some existing configurations and thus we cannot implement such change. Usually this problem was workarounded in modules via other means. One possibility is to use PAM data deallocation function to have some code run as late as possible in the pam evaluation. But I suppose this could be too fragile. Other possibility is to place the module twice in the session stack and mark the invocation with module option. Pam_selinux module uses this.

**Comment by msharov** [comment]
Not changing this already keeps existing configurations broken, requiring the workarounds you mention. Perhaps you could at least find out what configurations would break by doing things correctly. And then decide whether doing the right thing is better than lightly ruffling a few currently hypothetical feathers.

**Comment by msharov** [comment]
In fact, I am going to make the stronger statement that it impossible for any module to rely on the current behavior. No module can require another module to be accessible at session open while at the same time requiring that module to be inaccessible at session close. It is a logical impossibility. A module either needs its dependency or it does not. If it needs the dependency, then it will need it at both open and close, and will be broken by the current behavior.

Furthermore, anyone using the double-entry workarounds you mention will be unaffected by the fix, as the calling sequence will not change for them.

So I beseech you to reopen the bug, apply the fix, and end our misery!

**Comment by t8m** [comment]
The fix would not be simple at all. The evaluation of the stack is not trivial, there can be jumps, premature aborts of the evaluation, etc. The 'correct' or more properly called 'logical' implementation would have to create a list of modules called in the session-open call and then invoke them in reverse order in the session-close call. But that would definitely be an ABI or expected behavior breakage. It would make sense to do such change in some hypothetical Linux-PAM 2.0 release (and there would be other behavioral or ABI changes that would also make sense in such major release). Unfortunately the development of Linux PAM has recently fairly stalled.

So I will reopen but do not expect the fix to be in any Linux PAM release any time soon.

**Comment by msharov** [comment]
> The 'correct' or more properly called 'logical' implementation would 
> have to create a list of modules called in the session-open call and 
> then invoke them in reverse order in the session-close call.

No, that would not be correct. Session-open/close calls should not be treated like ctor/dtor pairs. Modules are not objects and can be called before session-open, such as a password manager module called in the auth section and then in session. A module may do something at session close only, or fail at open and still be ok. For example, pam_mount may fail at open due to the fs already being mounted, but should still be called at close to try unmounting it. So the current behavior of always calling all modules at close is still needed and reversing the order really is the only change necessary.

**Comment by t8m** [comment]
That's another good argument that what you propose also does not make sense.

And the conclusion might be that we actually need separate configuration stack for open and close session.

Simply reversing the stack will definitely not work in presence of jumps, immediate terminations of the stack, etc.

**Comment by t8m** [comment]
And that is also a good argument that PAM is in very hard need of serious redesign and making just isolated but intrusive changes like what you propose actually does not make sense.

**Comment by msharov** [comment]
> we actually need separate configuration stack for open and close session.

No, we don't. Modules can already easily decide whether to do work at open or close, or both. Most already do work at both. Adding more configuration options will only confuse the users. Having separate open and close stacks only in the code, as PAM does now, is already sufficient.

> Simply reversing the stack will definitely not work in presence of 
> jumps, immediate terminations of the stack, etc.

Module call ordering is determined by the configuration file, not the module code. AFAIK there is no way for a module to change its execution order (or pam_mount might have reordered itself to the end to fix its dependency problem), and there shouldn't be. Any module that tries to do that goes against the sysadmin's orders, and that is always wrong and will cause problems whether the stack is reversed or not.

As for terminations, session modules should not be allowed to do that. In the session list module failure does not affect other modules directly, since authentication is complete at that point and login proceeds anyway. Modules that terminate at session are simply badly written and their behavior would cause problems whether the close stack is reversed or not.

**Comment by t8m** [comment]
Do you understand that what you write here applies to only some possible configurations? Apparently not. Never mind, this is going nowhere and I certainly will not implement your proposed change.

**Comment by msharov** [comment]
Ok; it's your project and if you don't want the fix, you don't want the fix.

But you should at least explain to which configurations it would not apply. I know I am not the first or the last to complain about this issue, so clarifications would certainly help prevent future flamewars.

The way things are now, I am arguing to fix a problem affecting specific configurations that a lot of people use (pam_mount and systemd, for example); a problem that you yourself say that many people have to apply workarounds for. In opposition, you say that some configurations would be broken, but are unable to provide any real-world specific examples of such configurations so we could gauge how much harm, if any, the proposed change would do.

Again, some clarifications would be appreciated. If not for me, then for the multitudes of despondent users who will find this bug in the future.

**Comment by t8m** [comment]
Configurations using jumps in the stack.

**Comment by msharov** [comment]
I am sorry, but I don't understand. Could you please give a specific example of a configuration using jumps in the stack?

**Comment by msharov** [comment]
Sorry, nevermind. Jumps are certainly a highly undocumented feature. I have never seen it in any pam configuration file and have only been able to detect its existence by reading the source code.

From what I can see about the feature, it skips modules based on return code. I do not see how this can ever be used in the session stack. Session modules are all loaded together, there is never any reason to skip them, so I can not imagine any valid use for this there.

So how about this: after loading the configuration files, check the session close list for jumps and reverse if there aren't any? Document the behavior along with the jumps themselves and I think that would be an acceptable solution.

Alternatively, disallow jumps in session list altogether. I can not think of a single use case for this it undoubtedly qualifies as excessively programmable feature whose only purpose seems to be to make your life difficult. The detection can happen in the configuration parser and should be a pretty simple fix, adding one more condition at pam_handlers.c:188.

**Comment by Daniel-Abrecht** [comment]
Jumps are actually a pretty basic and useful feature. It's commonly used to do something depending on what the previous module did. I've seen it used to mount a directory only if a user successfully logs in.
I personally use it to log every failed login attemp on my server:
 - Live feed: https://preview.danielabrecht.ch/loginfails/
 - Logging program: https://github.com/Daniel-Abrecht/honeypot/blob/master/login_logger.c

One of my pam config files contains the following:
```
auth    [success=2 default=ignore]  pam_unix.so nullok_secure
auth    optional            pam_exec.so expose_authtok /usr/local/bin/login_logger /var/log/failed_logins
auth    requisite           pam_deny.so
auth    required            pam_permit.so
```

In other words, it skips pam_exec.so and pam_deny.so if the login (pam_unix.so) succeeds, otherwise it executes them. I think every debian/devuan system has those lines, except for the pam_exec line.

It's also pretty well documented in the pam.conf(5) manpage: http://man7.org/linux/man-pages/man5/pam.d.5.html
The relevant part is the following:
```
N (an unsigned integer)
           equivalent to ok with the side effect of jumping over the next N
           modules in the stack. Note that N equal to 0 is not allowed (and
           it would be identical to ok in such case).
```

Also, I recommend to just stop using systemd, it's not worth the hassle and just prevents people from learning how real systems actually work.

**Comment by msharov** [comment]
> Also, I recommend to just stop using systemd, it's not worth the hassle 
> and just prevents people from learning how real systems actually work.

Not a helpful suggestion! :) All major distributions use systemd and consequently so do most users. Sure I'd like to get rid of it, but it simply isn't practical.

> I've seen it used to mount a directory only if a user successfully logs in.

That is already correctly handled by pam_mount, there is no need to use jumps for it.

> I personally use it to log every failed login attemp on my server

That is what syslog is for.

But anyhow, I was asking for a jumps use for the session stack. I can see that there are possible uses in the auth stack, even though, as I point out above, these uses are better handled by other means. Unnecessary complexity such as this is, after all, one of the reasons people hate systemd.

The session stack is entirely different from the auth stack in that it is more like a systemd service supervisor for services requring login credentials. Thus its main use is to decrypt user's stuff, such as the home directory or the password database. Another use is for detecting logout events and cleaning up, such as what the systemd module does by stopping user services.

Systemd is on pretty much every major distribution. Many distributions, most notably Ubuntu, will ask you to encrypt your home directory during installation. So these uses of pam_systemd and pam_mount are on millions of systems, and they all suffer from the ordering problem in this bug. Conversely, jumps are a feature most people have never heard of, and whose capabilities are better implemented by other means, as I have pointed out above. So disabling it in the sessions stack while fixing the ordering would help millions and may cause inconvenience to a tiny number of people who would already be very experienced in working around it. I think it is a good tradeoff.

**Comment by Daniel-Abrecht** [comment]
I thought the jumps should be usable to change order by skipping it depending on the pam type, but it seams like only jumps at open_session are considered, and at close_session the module is skipped depending on if it was skipped at open_session, which seams strange to me.

I tried the following:
```
session [default=ok ignore=1] pam_exec.so type=open_session /bin/true
session optional pam_exec.so stdout /bin/sh -c [/bin/echo a $PAM_TYPE]
session optional pam_exec.so stdout /bin/sh -c [/bin/echo b $PAM_TYPE]
```

And I also tried this:
```
session [default=1 system_err=ignore] pam_exec.so quiet /bin/sh -c [test "$PAM_TYPE" != open_session]
session optional pam_exec.so stdout /bin/sh -c [/bin/echo a $PAM_TYPE]
session optional pam_exec.so stdout /bin/sh -c [/bin/echo b $PAM_TYPE]
```

I don't know why that doesn't work as I intendet...
